server/src/main/scala/org/http4s/server/ContextRouter.scala

```diff
@@ -16,15 +16,68 @@
 
 package org.http4s.server
 
+import cats.Contravariant
+import cats.Functor
+import cats.Monad
 import cats.data.Kleisli
+import cats.data.OptionT
 import cats.effect.Sync
-import cats.syntax.semigroupk._
+import cats.syntax.all._
 import org.http4s.ContextRequest
 import org.http4s.ContextRoutes
+import org.http4s.Request
 import org.http4s.Uri
 
 object ContextRouter {
 
+  object Segment {
+    implicit def functor[F[_]: Functor, X]: Functor[Segment[F, X, *]] =
+      new Functor[Segment[F, X, *]] {
+        override def map[A, B](fa: Segment[F, X, A])(f: A => B): Segment[F, X, B] =
+          new Segment(fa.run(_, _).map(f))
+      }
+
+    implicit def contravariant[F[_], X]: Contravariant[Segment[F, *, X]] =
+      new Contravariant[Segment[F, *, X]] {
+        override def contramap[A, B](fa: Segment[F, A, X])(f: B => A): Segment[F, B, X] =
+          new Segment((b, segment) => fa.run(f(b), segment))
+      }
+
+    class Partial[A] {
+      def apply[F[_], B](run: (A, Uri.Path.Segment) => OptionT[F, B]): Segment[F, A, B] =
+        new Segment(run)
+    }
+    // partially apply for inference
+    def apply[A] = new Partial[A]
+  }
+
+  final class Segment[F[_], A, B](val run: (A, Uri.Path.Segment) => OptionT[F, B]) {
+    def apply(routes: ContextRoutes[B, F])(implicit F: Monad[F]): ContextRoutes[A, F] =
+      Kleisli { case ContextRequest(a, req) =>
+        for {
+          head <- OptionT.fromOption[F](req.pathInfo.segments.headOption)
+          b <- run(a, head)
+          caret = req.attributes.lookup(Request.Keys.PathInfoCaret).getOrElse(0)
+          response <- routes(
+            ContextRequest(b, req.withAttribute(Request.Keys.PathInfoCaret, caret + 1))
+          )
+        } yield response
+      }
+
+    def ->(routes: ContextRoutes[B, F]): Routable[F, A] =
+      Routable.Dynamic(this, routes)
+  }
+
+  object Routable {
+    final case class Static[F[_], A](tupled: (String, ContextRoutes[A, F])) extends Routable[F, A]
+    final case class Dynamic[F[_], A, B](segment: Segment[F, A, B], routes: ContextRoutes[B, F])
+        extends Routable[F, A]
+
+    implicit def tuple[F[_], A](tupled: (String, ContextRoutes[A, F])): Routable[F, A] =
```

Comment:
If we delete this, would be that worsen ergonomics much? I'm just trying to stay away from implicits conversions when possible.

Reply:
Yes, If we remove that, then we would have a very different user experience.
I'm not a fan of adding implicit conversions, but in this case it provides a very smooth migration path from the existing static-only routers